File: clients/swift/AgentHostProtocol/Tests/AgentHostProtocolTests/FileEditTypesTests.swift
import Foundation
import XCTest
import AgentHostProtocol

final class FileEditTypesTests: XCTestCase {
    func testTypedFileEditsRoundTrip() throws {
        let side = FileEditSide(
            uri: "file:///workspace/file.txt",
            content: ContentRef(
                uri: "ahp-content:/file",
                sizeHint: 32,
                contentType: "text/plain",
                nonce: "v1"
            )
        )
        let edit = FileEdit(
            before: side,
            after: side,
            diff: FileEditDiffStats(added: 2_147_483_648, removed: 0)
        )
        let original = FileEditCollection(items: [edit])
        let encoded = try JSONEncoder().encode(original)
        let decoded = try JSONDecoder().decode(FileEditCollection.self, from: encoded)
        XCTAssertEqual(decoded.items.count, 1)
        let before: FileEditSide = try XCTUnwrap(decoded.items[0].before)
        let after: FileEditSide = try XCTUnwrap(decoded.items[0].after)
        let stats: FileEditDiffStats = try XCTUnwrap(decoded.items[0].diff)
        let added: Int? = stats.added
        XCTAssertEqual(before.uri, side.uri)
        XCTAssertEqual(after.content.uri, "ahp-content:/file")
        XCTAssertEqual(after.content.sizeHint, 32)
        XCTAssertEqual(after.content.contentType, "text/plain")
        XCTAssertEqual(after.content.nonce, "v1")
        XCTAssertEqual(added, 2_147_483_648)
        XCTAssertEqual(stats.removed, 0)
    }

    func testEmptyCollectionPreservesItems() throws {
        let encoded = try JSONEncoder().encode(FileEditCollection(items: []))
        let object = try XCTUnwrap(
            try JSONSerialization.jsonObject(with: encoded) as? [String: Any]
        )
        XCTAssertEqual(Set(object.keys), Set(["items"]))
        let items = try XCTUnwrap(object["items"] as? [Any])
        XCTAssertTrue(items.isEmpty)
    }
}

File: clients/dotnet/tests/AgentHostProtocol.Tests/FileEditTypesTests.cs
#nullable enable

using System;
using System.Collections.Generic;
using System.Text.Json;
using Microsoft.AgentHostProtocol;
using Xunit;

namespace Microsoft.AgentHostProtocol.Tests;

public sealed class FileEditTypesTests
{
    [Fact]
    public void TypedFileEditsUseGeneratedMetadata()
    {
        var options = new JsonSerializerOptions
        {
            PropertyNamingPolicy = JsonNamingPolicy.CamelCase,
            TypeInfoResolver = AhpJsonMetadata.Default,
        };
        foreach (Type type in new[]
        {
            typeof(FileEditSide),
            typeof(FileEditDiffStats),
            typeof(FileEditCollection),
        })
        {
            Assert.NotNull(AhpJsonMetadata.Default.GetTypeInfo(type, options));
        }

        var side = new FileEditSide
        {
            Uri = "file:///workspace/file.txt",
            Content = new ContentRef
            {
                Uri = "ahp-content:/file",
                SizeHint = 32,
                ContentType = "text/plain",
                Nonce = "v1",
            },
        };
        var original = new FileEditCollection
        {
            Items = new List<FileEdit>
            {
                new()
                {
                    Before = side,
                    After = side,
                    Diff = new FileEditDiffStats { Added = 2147483648L, Removed = 0 },
                },
            },
        };
        string wire = JsonSerializer.Serialize(original, options);
        var decoded = Assert.IsType<FileEditCollection>(
            JsonSerializer.Deserialize<FileEditCollection>(wire, options));
        FileEdit item = Assert.Single(decoded.Items);
        var before = Assert.IsType<FileEditSide>(item.Before);
        var after = Assert.IsType<FileEditSide>(item.After);
        var stats = Assert.IsType<FileEditDiffStats>(item.Diff);
        long? added = stats.Added;
        Assert.Equal(side.Uri, before.Uri);
        Assert.Equal("ahp-content:/file", after.Content.Uri);
        Assert.Equal(32L, after.Content.SizeHint);
        Assert.Equal("text/plain", after.Content.ContentType);
        Assert.Equal("v1", after.Content.Nonce);
        Assert.Equal(2147483648L, added);
        Assert.Equal(0L, stats.Removed);
    }

    [Fact]
    public void EmptyCollectionPreservesItems()
    {
        var empty = new FileEditCollection { Items = new List<FileEdit>() };
        string wire = SystemTextJsonAhpSerializer.Default.Serialize(empty);
        using JsonDocument document = JsonDocument.Parse(wire);
        JsonElement items = document.RootElement.GetProperty("items");
        Assert.Equal(JsonValueKind.Array, items.ValueKind);
        Assert.Equal(0, items.GetArrayLength());
    }
}
